DCP: Initial support for overlay blending

chadmed opened the pull request:
DCP can blend two surfaces, so we should wire this up.

We currently don't really have a valid use case for this functionality, however this will be required to make full use of zero-copy video decode once AVD is merged. Work is ongoing in userspace to properly make use of hardware overlay planes. Given DCP's provenance, the best example of this would be directly scanning out hardware-decoded video to the primary plane, then overlaying subtitles/media controls on a different plane.

Given that DCP dies if we try to blend more than two surfaces, we cannot expose an overlay plane _and_ a hardware cursor plane, as userspace has been designed to expect for the last 30 years. Even if we could, again owing to its provenance DCP cannot deal with the unique requirements of a cursor plane - it crashes if a framebuffer partially clips outside of screenspace. This can be hacked around and it _is_ possible to expose a mostly-working cursor plane, however Kwin seems to be the only compositor that is architecturally capable of gracefully handling this in the same way macOS does (fall back to software when required and go back to hardware once the driver says it's safe). As a part of bringing up support for overlay planes, Kwin will also be able to use a `DRM_PLANE_TYPE_OVERLAY` for a hardware cursor, negating the need for any hacks. Kwin will safely use a hardware cursor on our overlay plane, and compositors which can't/won't implement a similar feature will just use it as a traditional `DRM_PLANE_TYPE_OVERLAY`.

This series only implements support for ARGB/ABGR on the overlay plane. This is really only useful for cursors, media player overlays/graphics, etc. and does not magically make video work. The final piece in the zero-copy video playback puzzle will be supporting the direct scanout of planar YUV video formats. This may not be trivial.

alyssarosenzweig committed:
WIP: drm/apple: Add DCP display driver

Add a DRM/KMS driver for Apple system on chips using the DCP
coprocessor, namely the Apple M1. The DCP was added in Apple A14; this
driver does not apply to older iDevices.

This driver targets the DCP firmware API shipped by macOS 12.1.
Currently no incompatibilities with macOS 12.0.1 or 12.2.1 are known.

Signed-off-by: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Co-developed-by: Janne Grunau <j@jannau.net>
Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HACK: drm/apple: avoid DCP swaps without attached surfaces

Xorg startup with modesetting driver triggers this. Move vblank
signalling to dcp to avoid a circular dependency between apple_drv
and dcp.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Fix rt_bandwidth for t600x

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Add nop sr_set_uint_prop callback for t600x-dcp

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Reference only swapped out framebuffers

The framebuffer can be unreferenced by GEM while the display controller
is still using it for scanout resulting in IOVA faults and crashed dcp.
dcp has to hold a reference until the swap is complete.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Use "apple,asc-dram-mask" for rtkit iovas

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Relicense DCP driver as dual MIT / GPL v2.0

Link: https://oftc.irclog.whitequark.org/asahi-dev/2022-03-20#30747564
Link: https://oftc.irclog.whitequark.org/asahi-dev/2022-03-20#30747570

Signed-off-by: Alyssa Rosenzweig <alyssa@rosenzweig.io>
Signed-off-by: Janne Grunau <j@jannau.net>

alyssarosenzweig committed:
drm/apple: Implement suspend/resume for DCP

Use the firmware setPowerState callback.

Signed-off-by: Alyssa Rosenzweig <alyssa@rosenzweig.io>

jannau committed:
drm/apple: dcp: fix TRAMPOLINE_IN macro

fixup! WIP: drm/apple: Add DCP display driver

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: implement read_edt_data

Handling it with trampoline_false can result in errors due to
uninitialized data.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: clear callback's output data

If there is a mismatch in the output size this way we have at leas
consistant results.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Support memory unmapping/freeing

Used by dcp on mode changes on the Macbook Pro 14" (M1 Max).

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
WIP: drm/apple: Change the way to clear unused surfaces

This seems to be incorrect but the flag seems to be related to clearing.

Allows unmapping surfaces after the swap finished.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: toggle power only when active state changes

squash! WIP: GPU/apple: t600x work in progress

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Start coprocessor on probe

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Switch to nonblocking commit handling

The swap completes only after the async reply from DCP. Uses
drm_atomic_helper_wait_for_flip_done instead of
drm_atomic_helper_wait_for_vblanks. This should allow ius to get rid
of the scheduled fake vblanks.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Log callbacks with their tag as debug output

Mostly for the generic callbacks nop, true, false.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Add DCP interface definitions used on t600x

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Use a device tree defined clock for dcpep_cb_get_frequency

Frequency differs between M1 and M1 Pro/Max/Ultra.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Clear used callback/cookie on dcp_ack

Avoids unexpected callbacks on nesting state errors. Encountered when
making DCP calls from a second thread for the backlight.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Add t600x support

Call power-on/-off handling explicitly from apple_crtc_atomic_enable /
apple_crtc_atomic_disable. This makes DPMS work.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Add somewhat useful debug prints

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Add less tons of questionable debug prints

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: laod piodma dev via explicit phandle

Use a device_link to ensure piodma has is bound to its DART.

fixup! WIP: drm/apple: Add DCP display driver

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/apple: Fix kzalloc in dcp_flush()

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
drm/apple: Allow modesets even when disconnected

Fixes a display wakeup issue seen with Plasma/X11.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Mark the connecter on init only with modes as connected

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: make note about drm.mode_config.max_width/height

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Split dcpep/iomfb out of dcp.c

For external display support DCP will use more endpoints.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
WIP: add header test target copied from i915

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Fix connector state on devices with integrated display

DCP issues hotplug_gated callbacks after SetPowerState() calls on devices
with display (macbooks, imacs). This must not result in connector state
changes on DRM side. Weston will not re-enable the CRTC after DPMS off
if the connector is not in connected state.
DCP provides with dcp_is_main_display() a call to query if the device has
an integrated display.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Replace atomic refcount with kref

fixup! drm/apple: Add t600x support

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Start using tracepoints

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Unbreak multiple DCP plane <-> crtc matching

Still untested so this changes the status from definitively broken to
probably broken.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Add support for DRM_FORMAT_XRGB2101010

iboot uses DRM_FORMAT_XRGB2101010 at boot announce preference by listing
it as first format.

Disabled for now since it results in oversaturated colors. Might be
fixed by using "IOMFB::UPPipe2::set_matrix()".

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Add apple_drm_gem_dumb_create()

DCP needs a 64-byte aligned pitch, override drm_gem_dma_dumb_create()
to align the pitch as necessary and use
drm_gem_dma_dumb_create_internal() to allocate the GEM object.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Reject modes without valid color mode

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Convert 2 non-assert WARN()s to dev_err()

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Send an disconnected hotplug event on ASC crash

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Add dcp_crtc_atomic_check

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Fix DCP run time PM

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Fix DCP initialisation

Try to avoid races between DRM driver and DCP(ext) probing and
initialisation:
- do not start RTKit endpoints in dcp probe, iomfb excepts DRM objects
  to be already initialized on startup
- ensure in apple_drv that all DCPs are probe via device links
- initialize DRM planes, connectors, encoders
- start DCP RTKit endpoints synchronously

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Specify correct number of DCP*s for drm_vblank_init

Unbreaks dcpext a little further.

fixup! WIP: drm/apple: Add DCP display driver

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Use connector types from devicetree

Needs to be re-done for upstream submission but the exisiting port,
bridge, connector and display bindings are a bad fit since DCP manages
everything.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Remove other framebuffers before DRM setup

Allows taking over the device node from simpledrm and appaers to be the
common pattern in DRM drivers replacing boot framebuffers.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Provide notch-less modes

If the device tree caries a "apple,notch-height" property subtract it
from all modes and render all framebuffers offsetted by it.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Support opaque pixel formats

Opaque pixel formats such as XRGB8888 or YUV formats require flag in
struct dcp_surface to be set to be displayed properly.
Fixes display issues with fbcon after the handover from simpledrm on
j314c with 16:10 modes (notch hiding).

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: iomfb: Use FIELD_{GET,PREP}

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: iomfb: Unify call and callback channels

AP calls initiated inside callbacks are using the callback channel and
context.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: "match" PMU/backlight services on init

Verify that this still works on HDMI/USB-C.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Brightness control via atomic commits

This abuses color_mgnt_change in drm_crtc_state and will be changed once
phase 2 of the "drm/kms: control display brightness through drm_connector
properties" RfC (linked below) is implemented.

The lookup of DAC values from brightness (nits) is not fully understood.
Since IOMFB reports te brightness back the easiest solution would be to
create our own lookup table or find a approximation which works.

DCP appears to report the brightness in nits by
"PropRelay::pr_publish(prop_id=15, value=...)" (scaled by
"Brightness_scale").

Link: https://lore.kernel.org/dri-devel/b61d3eeb-6213-afac-2e70-7b9791c86d2e@redhat.com/
Signed-off-by: Janne Grunau <j@jannau.net>

svenpeter42 committed:
mux: apple DP xbar: Add Apple silicon DisplayPort crossbar

This drivers adds support for the display crossbar used to route
display controller streams to the three different modes
(DP AltMode, USB4 Tunnel #0/#1) of the Type-C ports.

Signed-off-by: Sven Peter <sven@svenpeter.dev>

svenpeter42 committed:
drm: apple: Add oob hotplug event

Signed-off-by: Sven Peter <sven@svenpeter.dev>

svenpeter42 committed:
drm: apple: Add DPTX support

This is required for DP Altmode, DP Thunderbolt tunneling and HDMI
output on 14/16-inch Macbook Pros and M2* desktop devices.

M2* desktops and 14 and 16 inch Macbook Pros expose a DisplayPort to
HDMI converter which is driven by the DP output of one of the DCP/DCPext
display coprocessor/controller blocks.
Two gpio pins are used for power control. Another gpio pin acts as HDMI
hpd.
Do not use the hpd as direct drm_connector interrupt since that is
already wired to DCPs hotplug notification. Instead use it to trigger
link setup via the dptx endpoint.

Signed-off-by: Sven Peter <sven@svenpeter.dev>
Co-developed-by: Janne Grunau <j@jannau.net>
Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Fix shutdown of partially probed dcp

No need to shut the co-processor down if it wasn't booted to begin with.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Set maximal framebuffer size correctly

DCP reports this in the IOMFBMaxSrcPixels dictionary. Use that instead
of the hardcoded values from M1 Max.
Fixes multiscreen X11 setups.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Prevent NULL pointer in dcp_hotplug

A bit hackish, probably missing something in the setup or simply calling
this too early.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Update date last update

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
HACK: gpu: drm: apple: j314/j316: Ignore 120 Hz mode for integrated display

It's currently not useful as DCP limits the swap rate to 60 Hz anyway
and marcan reported pointer choppiness with 120 Hz.

Signed-off-by: Janne Grunau <j@jannau.net>

marcan committed:
drm/apple: Fix suspend/resume handling

Use the drm_modeset helpers for suspend/resume at the subsystem level,
which actually do the proper save/restore dance and work, instead of
open-coding calls to dcp_poweroff/dcp_poweron which is clearly wrong and
doesn't restore properly (nor would it be correct if the display was
already off when suspended).

Also fix apple_platform_remove while I'm here, since drvdata wasn't
getting set so that would never work.

Signed-off-by: Hector Martin <marcan@marcan.st>

jannau committed:
gpu: drm: apple: Avoid drm_fb_dma_get_gem_addr

It adjust the address by the source position duplicating setting the
source postion in IOMFB's swap_submit struct. Prefer the later since it
is more explicit.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: register backlight device after IOMFB start

This allows us to specify the boot display brightness as initial
brightness of the baclight device.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Add trace point for display brightness

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Implement drm_crtc_helper_funcs.mode_fixup

Prevents KDE for now to set modes not reported by IOMFB. Seen on j493
with the common display resolution of 2560x1600.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Read display dimensions from devicetree

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/apple: Disable fake vblank IRQ machinery

The hardware does not have a vblank IRQ and drm already knows how to
deal with that appropriately, so don't pretend it does.

Fixes Xorg.

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/apple: Wait for power on request to complete synchronously

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/apple: Remove obsolete ignore_swap_complete

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/asahi: Fix backlight restores on non-microLED devices

Apparently what happens here is that the DCP's idea of backlight
brightness is desynced with the real brightness across power cycles.
This means that even if we just force an update after a power cycle, it
doesn't work since it considers it unchanged. To fix this, we need to
both force an update on poweron and also explicitly turn the backlight
off on poweroff, which makes DCP listen to us and actually update the
backlight state properly.

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
drm/apple: Schedule backlight update on enable_backlight_message_ap_gated

On non mini-LED displays the backlight comes out of power-off (DPMS)
with minimal backlight brightness. This seems to be a DCP firmware
issue. It logs "[BrightnessLCD.cpp:743][AFK]nitsToDBV: iDAC out of
range" to syslog although the brightness in the swap_submit call is
valid.

This fixes the issue only for clients using swap. For other clients an
atomic backlight update has to be scheduled via a work queue.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Report "PMUS.Temperature" only for mini-LED backlights

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Check if DCP firmware is supported

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Parse color modes completely

Selecting the mode with the highest score may result in HDR mode for
some displays. Since HDR is not support on driver side this produces
an unexpected color representation.
Full parsing allows us to reject virtual color modes (should already be
invalid due to missing "Score").
Preparation for color and timing mode tracing.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Add tracing for color and timing modes

Restrict symbol mapping for EOTF, pixel encoding and colorimetry to
tracing due to low confidence that it is correct. Main concern is the
colorimetry mapping.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Skip parsing elements of virtual timing modes

Prevents trace points of unused color modes.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Prefer SDR color modes

DCP best scored color mode might result in an HDR mode. As long as the
driver (and DRM) is not ready for HDR try to avoid such modes.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: reenable support for {A,X}RGB2101010

Seems to work now with 'surface.colorspace = 12'.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Add IOMobileFramebufferAP::get_color_remap_mode

Probably not important but avoids an unnecessary difference compred to
macOS.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Add show_notch module parameter

Can be used on devices with camera notch to use the full display height
and thus show the notch.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
Revert "gpu: drm: apple: reenable support for {A,X}RGB2101010"

This reverts commit d39542179a8f5a5d2e80eebf77bc739857a1051c.

'w30r' is a wide gammut mode. As long as the display is SDR DCP will end
up displaying picture correctly but on HDR displays like the display in
the Macbook Pro 14"/16" (2021) or external displays with HDR EOTF the
picture has oversaturated / black colors.

The non-wide gammut 10-bit per component pixelformats "l10r" and "R10k"
 are not supported.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Use drm_aperture_remove_conflicting_framebuffers

This does not seem to be as racy as drm_aperture_remove_framebuffers()
and seems to reliably takes over simpledrm's device node.

Signed-off-by: Janne Grunau <j@jannau.net>

marcan committed:
drm/apple: Enable 10-bit mode & set colorspace to native

This works on both 8-bit and 10-bit modes without any weirdness, and
gives us the native colorspace without any conversion. Color correction
should probably be handled in software anyway.

However, we need to use surface 1 (at least on t600x), since 0 seems
stuck in bg-sRGB mode for some reason...

Signed-off-by: Hector Martin <marcan@marcan.st>

jannau committed:
gpu: drm: apple: Use components to avoid deferred probing

There was a report of a race between DRM device registration (and
removal of the simpledrm device) and GDM startup.

The component based device binding ensures that all necessary devices
are bind in the probe method of the last missing component.

Technically the piodma-mapper should be a component of dcp but since it
is only used for its iommu it can be a component of the display
subsystem.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Wait for iomfb initialization

Avoids "[drm] Cannot find any crtc or sizes" during fbdev initialization
if a display is connected.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
gpu: drm: apple: Clear all surfaces on startup

With "drm/apple: Enable 10-bit mode & set colorspace to native" kernel
log messages are shown in an Apple logo shaped region in the middle of
the display when using BGRA.
The field currently identified as "opaque" is mislabeled and has to be
investigated further.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Use drm_module_platform_driver

This check for the "nomodeset" kernel command line parameter in its
register method.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Allocate drm objects according to drm's expectations

drm's documentaion explicitly tells us not to use devm_kzalloc(). drm
device structures might out live the device when they are in use by
userspace while the device vanishes.

jannau committed:
drm/apple: Update swap handling

- opaque -> is_premultiplied
- swap_enabled BIT(31) seems to be update background with
  dcp_swap.bg_color
- add unused fields is_tearing_allowed, ycbcr_matrix, protection_opts,
  unk_num, unk_denom

Changes: use is_premultiplied only for XRGB8/XBGR8, Update background
only when necessary.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: simplify IOMFB_THUNK_INOUT

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Add callbacks triggered by last_client_close_dcp()

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Use backlight_get_brightness()

Backlight drivers are expected to use this instead of accessing
backlight properties.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Set backlight level indirectly if no mode is set

Fixes following warning when systemd-backlight restores the backlight
level on boot before a mode is set:

Call trace:
  drm_atomic_helper_crtc_duplicate_state+0x58/0x74
  drm_atomic_get_crtc_state+0x84/0x120
  dcp_set_brightness+0xd8/0x21c [apple_dcp]
  backlight_device_set_brightness+0x78/0x130
  ...

Signed-off-by: Janne Grunau <j@jannau.net>

svenpeter42 committed:
drm: apple: DCP AFK/EPIC support

Signed-off-by: Sven Peter <sven@svenpeter.dev>
Co-developed-by: Martin Povišer <povik+lin@cutebit.org>
Signed-off-by: Martin Povišer <povik+lin@cutebit.org>
Co-developed-by: Janne Grunau <j@jannau.net>
Signed-off-by: Janne Grunau <j@jannau.net>

povik committed:
gpu: drm: apple: Expose injecting of EPIC calls via debugfs

Signed-off-by: Martin Povišer <povik+lin@cutebit.org>
Co-developed-by: Janne Grunau <j@jannau.net>
Signed-off-by: Janne Grunau <j@jannau.net>

povik committed:
gpu: drm: apple: Set up client of AV endpoint

Signed-off-by: Martin Povišer <povik+lin@cutebit.org>

hoshinolina committed:
drm/apple: Fix bad error return

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/apple: Fix parse_string() memory leaks

Signed-off-by: Asahi Lina <lina@asahilina.net>

povik committed:
gpu: drm: apple: Add utility functions for matching on dict keys

Signed-off-by: Martin Povišer <povik+lin@cutebit.org>

povik committed:
gpu: drm: apple: Add sound mode parsing

Signed-off-by: Martin Povišer <povik+lin@cutebit.org>

jannau committed:
drm/apple: Add support for the macOS 13.2 DCP firmware

This adds support for multiple incompatible DCP firmware versions.
The approach taken here duplicates more than necessary. Unmodified calls
do not need to be templated. For simplicity and in the expectation that
more calls and callbacks are modified in the future everything is
templated.

Signed-off-by: Janne Grunau <j@jannau.net>

povik committed:
ALSA: Introduce 'snd_interval_rate_bits'

Signed-off-by: Martin Povišer <povik+lin@cutebit.org>

povik committed:
gpu: drm: apple: Add DCP audio driver

Signed-off-by: Martin Povišer <povik+lin@cutebit.org>

povik committed:
gpu: drm: apple: Add 'parse_blob'

Signed-off-by: Martin Povišer <povik+lin@cutebit.org>

jannau committed:
drm/apple: purge unused dcp_update_notify_clients_dcp

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: ignore surf[3] in clear swap calls

MacOS 13.2 does the same and it is unclear if surf[3] can be used at
all. PRobably not necessary but found during debugging to firmware 13.2.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Support color transformation matrices

kwin 5.27.3 adds support for "Night Color" via drm "CTM" properties.
Wire CTM support up via the "set_matrix" iomfb call.

Link: https://bugs.kde.org/show_bug.cgi?id=455720
Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/apple: Align buffers to 16K page size

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
drm/apple: Drop unsupported DRM_FORMAT_ARGB2101010

Depends on https://gitlab.freedesktop.org/asahi/mesa/-/merge_requests/5

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Move panel options to its own sub-struct

Fixes overwriting the panel's physical dimensions on poweroff/sleep.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: afk: Adapt to macOS 13.3 firmware

Signed-off-by: Janne Grunau <j@jannau.net>

marcan committed:
dcp: Allow unused trampolines

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
dcp: Add get_tiling_state

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
dcp: 42-bit DMA masks

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
dcp: T602X bwreq support

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
dcp: Warn if DMA mapping fails

Signed-off-by: Hector Martin <marcan@marcan.st>

jannau committed:
WIP: drm/apple: Port to incompatible V13.3 firmware interface

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Remove simpledrm framebuffer before DRM device alloc

Should result in drm apple to be registered as first DRM device
replacing simpledrm. Should resolve problems with userspace assuming
that card0 is the main displays device.

Signed-off-by: Janne Grunau <j@jannau.net>

marcan committed:
drm/apple: Mark DCP as being in the wakeup path

This prevents the PD from being shut down on suspend, which we need
until we support runtime PM properly again.

Signed-off-by: Hector Martin <marcan@marcan.st>

jannau committed:
drm: apple: dptx: Port APCALL to macOS 13.3 firmware

The 13.3 firmware has an additional get_max_lane_count call inserted
with ID 10.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
mux: apple dp crossbar: Support t8112 varient

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
mux: apple dp crossbar: FIFO_RD_UNK_EN seems to use 2 bits per dispext*

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
mux: apple dp crossbar: Read UNK_TUNABLE before and after writing it

Makes traces easier to compare with macOS.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Only match backlight service on DCP with panel

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: implement abort_swaps_dcp

To match macOS behavior and in the hope to fix dcpext crashes on t8112.
Crashes still occur but let's keep this. Shouldn;t make a difference
since we're on the swaps to finish.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: limit backlight updates to integrated panels

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: Increase modeset timeout to 2.5 seconds

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Get rid of the piodma dummy driver

It's only needed to configure the display contoller's iommu to share
buffers between the DCP co-processor and the display controller.
Possible concern is runtime PM for it and its iommu. If we don't set it
up the power domain might never go to lower power states even if it
could.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: backlight: avoid updating the brightness with a commit

An atomic_commit for brightness changes will consume a DCP swap without
frame buffer updates and will result in a lost frame. After updating
the next brightness values wait for 1 frame duration (at 23.976 fps).
Check if the brightness update still needs to be send to DVCP or if a
swap did that in the meintime.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm/apple: Use iommu domain for piodma maps

The current use of of dma_get_sgtable/dma_map_sgtable is deemed unsafe.
Replace it with an unmanaged iommu domain for the piodma iommu to map
the buffers.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Align PIODMA buffers to SZ_16K

The iommu scatter table/list mapping can only map full iommu page size
extents. Just align the actual the allocation to the iommu page size.
This could be handled differently using DARTs subpage protection but
there's no easy way to integrate that.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
mux: apple dp crossbar: Support t602x DP cross bar variant

This is a simplified version and probably should live in a separate
file. Even the shared registers are quite different.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: Do not match/create PMU service for dcpext

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: port interface to macOS 13.5 firmware

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Add D129 allocate_bandwidth iomfb callback

Used on M2 Ultra During startup. Units are unclear.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Update supported firmware versions to 12.3 and 13.5

Removes support for all firmware versions which report as compatible to
13.3 except 13.5. This will be removed after m1n1 reports firmware 13.5
as "apple,firmware-compat" for a while.
The files with "v13_3" will be renamed at a later point to avoid
conflicts with development trees.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Remove explicit asc-dram-mask handling

This is no longer necessary after introducing "apple,dma-range" for the
dart driver.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: afk: Update read pointer before processing message

Avoids out of order messages and already unmapped buffers while tracing
with hv/trace_dcp.py.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: Increase modeset tiemout to 8.5 seconds

DCP itself uses with the 13.5 firmware a timeout of 8 seconds for
modesets. Using a longer timeout prevents overlapping calls to dcp and
might improve reliabilty with slower displays.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: Add set_active_lanes APCALL

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: Add DPTX_APCALL_ACTIVATE

Configures the phy to the correct dcp(ext) source by abusing submode in
the phy_set_mode_ext() call.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: Adapt dptxport_connect() to observed behavior

Adapt to behavior seen on j474s with dcp0 driving lpdptx-phy and
dp2hdmi using the macOS 13.5 firmware.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dcp: Port over to DEFINE_SIMPLE_DEV_PM_OPS

Avoids ugly "__maybe_unused".

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dcp: Remove cargo-culted devm_of_platform_populate

It does not do anything for dcp and its iommu only child node.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Move offsets for rt_bandwidth callback to DT

The offsets differ for every DCP instance. Instead of hardcoding offsets
for each SoC family offsets and calculate the instance offset move
everything to the device tree. This helps multi die SoCs since there is
and unexpected offset between both dies.
On multi die SoCs device tree changes were necessary to avoid
translating the PMGR reg via the seconds die "ranges" property.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: afk: Use linear array of services

"Channel numbers" as received by AFK/EPIC are constantly increasing over
restarts of the endpoint. Use a linear array of services and match based
on the channel number. The number of services per endpoint is too small
to make a difference.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: av: Support macOS 12.3 and 13.5 firmware APIs

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: av: Do not open AV service from afk receive handler

Use a completion to do it from avep_init() instead.

Signed-off-by: Janne Grunau <j@jannau.net>

marcan committed:
drm: apple: afk: Clear commands before sending them

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
drm: apple: Fix missing unlock path in dcp_dptx_connect

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
drm: apple: dptxep: Fix reply size check

Signed-off-by: Hector Martin <marcan@marcan.st>

marcan committed:
drm: apple: dptxep: Implement drive settings stuff

Just in case, for consistency with macOS.

Signed-off-by: Hector Martin <marcan@marcan.st>

jannau committed:
drm: apple: Implement D592 callback

This callback is occasionally seen around (failed) modesets. There seems
to be no need to handle it so just trace it.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Keep information at which swap_id fb are still referenced

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
Revert "drm: apple: iomfb: Do not match/create PMU service for dcpext"

This reverts commit ab69434d230f9951644e10c9142dbc43ea0516c4.

jannau committed:
drm: apple: HACK: Do not delete piodma platform device

of_platform_device_destroy() can trigger several NULL pointer
dereference which have been elusive so far.
Comment this for now since the oopses causes the shutdown to hang.
Since dcp can not be reloaded this leaks the platform device on shutdown
and reboot.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: Implement APCALL_DEACTIVATE and reset the phy

This mirrors what macOS does and should make reconnections more
reliable.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Disconnect dptx When the CRTC is powered down

Seems to make disconnect / reconnect more reliable and almost fixes
suspend/resume. The drm device tries to modeset too early on resume
which leaves the screen blank.
This should reduce power consumption after disconnecting the HDMI port.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: Wait for completion of dptx_connect.

Makes connects more reliable.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: HPD: Only act on connect IRQs

DCP notices the disconnects on its own and the parallel handling just
results in confusion (both on DRM and developer side).

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: Improve hotplug related logging

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/apple: Add missing sound Kconfig dependencies

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
drm: apple: Extract modeset crtc's atomic_flush()

Triggering modesets from drm_connector_helper_funcs.atomic_check is more
in line with DRM/KMS' design and allows returning errors from failed
modesets.
Ignore hotplug callbacks from DCP during modeset. DCP always does
disconnected -> connected on (at least the initial) modeset. Shield drm
helpers from this.
This improves reliability with externel (dptx based) displays.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: Log connect/disconnect calls

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Move modeset into drm_crtc's atomic_enable

squash! drm: apple: Extract modeset crtc's atomic_flush()
Fixes: 99d7bb861908 ("drm: apple: Extract modeset crtc's atomic_flush()")
Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Fix DPTX hotplug handling

- Do not trigger an hotplug event from disconnect. DCP/iomfb notices that
  itself.
- Check HPD status before disconnecting DPTX in the crtc disable
  path.
- disconnect on suspend to allow an orderly re-connect on resume

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: Use drm_kms_helper_connector_hotplug_event

Avoid device wide hotplugs as DCP knowns the affected connector.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm : apple: iomfb: Handle OOB ASYNC/CB context

Only observed with dcp/dptx in linux after initialisation and reset in
m1n1. On the initial startup dcp sends two D576 (hotPlug_notify_gated)
presumendly due to state confusion due to  the multiple dptx
connections.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: Extend hotplug/mode parsing logging

Under unknown but slightly broken conditions dcp sends timing modes
without linked color modes. Log a warning when this happens and log the
number of valid modes before emitting HPD events.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Adjust startup sequence and timing for dptx

DPTX setup from an initialized connection and display with sleeping and
reset dcp is unfortunately quite fragile. The display connection has to
be stopped and reestablished. Goodbye flicker free boot.
If the IOMFB endpoint is started too early dcp might provide incomplete
timing modes which prevent modesets.
On display standby a HPD is triggered should result in a fully
initialized dcp. If not a display cable unplug and plug should help.
MacOS doesn't handle this at all and just gives up.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dcp: Fix resume with DPTX based display outputs

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Prefer RGB SDR modes

DCP color mode scoring seems to prefer high bit depth color modes even
when it it would require DSC. For example 12-bit 4k 60 Hz YCbCr 4:4:4
over a 600 MHz HDMI 2.0 link.
Prefer 8-/10-bit RGB or YCbCr 4:4:4 modes if available.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Be less noisy about teardown notifies without service

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Add dcpav-service-ep

Known uses EDID retrieval and raw I2C access.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: Wait for link config on connect

Should make connect more reliable by avoiding hardcoded waits which are
either to long or too short. In the second case the display can't be
brought up since dcp fails to report any modes during start.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: Always parse DisplayAttributes

Fixes missing physical display dimensions for HDMI display on Macbook
Pros.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: parser: constify parser data

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: epic: Pass full notfiy/report payload to handler

The payload is not necessarily epic_std_service_ap_call. The powerlog
service on the system endpoint passes serialized dictionaries as
payload.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: epic: systemep: Parse "mNits" log events

The 13.5 firmware has stopped updating the NITS property on backlight
brightness changes. Parse system log events instead which report
backlight's brightness in millinits.
Fixes the backlight device's "actual_brightness" property used by the
systemd backlight service to save and restore brightness.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: export property dicts in connector debugfs

Signed-off-by: Janne Grunau <j@jannau.net>

kettenis committed:
drm: apple: backlight: force backlight update after resume

If the DCP firmware indicates that it didn't restore the
brightness, schedule an update.  Wait for 1 frame duration and
check if the brightness update has been taken care of by a swap
that happened in the meantime.
Fixes restoring the brightness after resume when running on a
dumb framebuffer where swaps may not happen for a very long time.

Signed-off-by: Mark Kettenis <kettenis@openbsd.org>

alyssais committed:
drm/apple: Add missing RTKit Kconfig dependency

Signed-off-by: Alyssa Ross <hi@alyssa.is>

arndb committed:
drm: apple: mark local functions static

With linux-6.8, the kernel warns about functions that have no
extern declaration, so mark both of these static.

Fixes: 2d782b0d007d ("gpu: drm: apple: Add sound mode parsing")
Signed-off-by: Arnd Bergmann <arnd@arndb.de>

jannau committed:
drm: apple: Add Kconfig option for audio

Signed-off-by: Janne Grunau <j@jannau.net>

jonathangray committed:
drm/apple: spelling fixes

Signed-off-by: Jonathan Gray <jsg@jsg.id.au>

jannau committed:
drm: apple: Fix/remove log messages

Add missing training '\n' and remove leftover dev_dbg() statements.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: Debounce HPD by simple msleep()

Not necessarily only a debounce but 500ms sleep in the HPD interrupt
handler seems to make the modeset more reliable on M2* desktop devices.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptx: Remove DPTX disconnect/connect on init

This was only necessary for dcp0 on M2* devices presumably because the
reset in m1n1 doesn't work as intended.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: audio: init AV endpoint later

This seems to get rid of initialization timeouts / failures.

Signed-off-by: Janne Grunau <j@jannau.net>

cschutijser committed:
drm: apple: backlight: release lock in error path

Signed-off-by: Caspar Schutijser <caspar@schutijser.com>

jannau committed:
drm: apple: av: Use a workqueue

Functionally a revert of
"drm: apple: av: Do not open AV service from afk receive handler"
with more workqueues.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: audio: move the audio driver into the DCP module

Those two drivers are closely linked and should always exists together.

Signed-off-by: Janne Grunau <j@jannau.net>

jonathangray committed:
drm/apple: fix double words in comments

Signed-off-by: Jonathan Gray <jsg@jsg.id.au>

jannau committed:
drm: apple: audio: Make the DP/HDMI audio driver a full driver

The main advantage is that it allows runtime PM which would have been
manually implemented with the ad-hoc instantiated platform driver.
This also probes the devices as component of the DRM driver which allows
to simplify the the interface between the av endpoint and the audio
driver.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: audio: Avoid probe errors

Now that the DP audio driver is a component of the display sub-system
probe errors will bring down the whole display initialization.
To prevent that the audio driver must not fail. Allow delayed sound card
initialization if the DMA controller is not ready, for example because
the apple-sio module is missing (at all or just in the initeramfs).
In the case apple-sio is available later provide as sysfs file
"probe_snd_card" to trigger initialization.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Switch back to drm_atomic_helper_commit_tail_rpm()

The custom commit_tail implementation stopped making after "drm/apple:
Disable fake vblank IRQ machinery" which stopped calling
drm_vblank_init(). Revert back to the standard helper implementation.
Avoids or at least significantly reduces page flips taking approximately
one frame time in kwin_wayland 6.

Fixes: ("drm/apple: Switch to nonblocking commit handling")
Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Fix broken MemDescRelay::release_descriptor callback number

Two callbacks for IOMFB::MemDescRelay seems to be dropped between 12.3
and 13.5 DCP firmware. This results in the renumbering of
MemDescRelay::release_descriptor from D456 to D454.
Noticed while when switching the display refresh rate to 50 Hz with a
14.5 system firmware on a M1 Max Macbook Pro.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: Reduce log spam about busy command channel

The most likely cause for this is an unexpected callback form which the
current driver doesn't recover. Warn only once about it.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: av: Warn only once about failed calls

Reduce log spam while errors are still likely due missing state checks.

jannau committed:
drm: apple: disable HDMI audio by default

Can be still enabled by adding `apple_dcp.hdmi_audio` the kernel command
line.

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/apple: Fix missing mode init (feel free to squash)

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau committed:
drm: apple: Override drm_vblank's page flip event handling [HACK]

Since we don't init/uses drm's vblank support our page flip timestamps
are CLOCK_MONOTONIC timestamps during the event generation. Since
compositors use the timestamp to schedule their next kms commit this is
timing sensitive sop move it under the drivers control.
Take the timestamp directly in the swap_complete callback.
Framebuffer swaps are unfortunately not fast with DCP. Measured time
from swap_submit to swap_complete is ~1.5 ms for dcp and ~2.3 ms for
dcpext. This warrants further investigation. Presentation timestamps
might help if delay on dcp firmware side occurs after the actual swap.
In the meantime doctor the time stamps and move the page flip completion
up to 1 ms earler. This fixes half rate refresh on external displays
displays using dcpext.

Signed-off-by: Janne Grunau <j@jannau.net>

chadmed committed:
drm: apple: respect drm_plane_state zpos

The for_each_oldnew_plane_in_state iterator is nondeterministic
in terms of the order of planes. DCP expects surfaces to be
fed to it in the correct order. Relying on the iterator to
lazily increment the index into surf[] means we cannot meet
this expectation. The constant reordering of planes in
the surf[] array seems to cause DCP to crash under certain
circumstances. Cursors will also often be drawn under the main
plane, which is less than ideal.

Populate surf[] in the order everyone expects us to. This fixes
a whole host of odd behaviour when wiring up multiple DRM universal
planes.

Signed-off-by: James Calligeros <jcalligeros99@gmail.com>

chadmed committed:
drm: apple: constrain swaps to maximum blendable surfaces

Despite having 4 surfaces, DCP can only blend two of them at once.

Constrain swaps to two surfaces, and warn if userspace somehow
tries to give us more to swap.

Signed-off-by: James Calligeros <jcalligeros99@gmail.com>

chadmed committed:
drm: apple: reject plane commit if it will crash DCP

Owing to its origin in mobile devices and the Apple TV, DCP seems to
have been designed under the assumption that no one could possibly want
a rectangle to clip the screen. If a rectangle's bottom-right edge
clips the screen, DCP will instead try to scale the destination rectangle
to the best of its ability... until it can't anymore.

DCP is not tolerant to faults and will crash if the onscreen portion of
the framebuffer ends up smaller than 32x32, or if any dimension ends up
entirely offscreen.

Use apple_plane_atomic_check() to reject requested plane states that
could crash DCP. This is the final piece of the puzzle required to
enable preliminary support for overlay planes on Apple Silicon devices.

Signed-off-by: James Calligeros <jcalligeros99@gmail.com>

chadmed committed:
drm: apple: add support for overlay planes

DCP is capable of compositing two surfaces in hardware. This is
important for zero-copy video playback, etc.

Set up an overlay plane so that userspace can do cool things with
it.

Signed-off-by: James Calligeros <jcalligeros99@gmail.com>

chadmed committed:
drm: apple: use correct min/max plane scaling factors

Fix the call to drm_atomic_helper_check_plane_state to use
the correct scaling factors.

Signed-off-by: James Calligeros <jcalligeros99@gmail.com>

chadmed committed:
drm: apple: warn about broken sw cursor fallback

Some userspace may not handle invalid plane checks gracefully
when falling back to a software cursor. This will manifest
as the screen freezing, recoverable by moving the cursor away
from a screen edge. Throw a warning once to let the user know
why this has happened.

Signed-off-by: James Calligeros <jcalligeros99@gmail.com>

jannau committed:
drm: apple: dptx: Fix get_drive_settings retcode

This appears to be lane count as "2" is observed for USB-C DP alt mode
in shared DP/USB3 mode.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: dptxport: get_max_lane_count: Retrieve lane count from phy

This unfortunately doesn't work relieably with typec-altmode-displayport
since the oob hotplug notification arrives before atc-phy is configured
to the appropiate DP mode.

Signed-off-by: Janne Grunau <j@jannau.net>

hoshinolina committed:
drm/apple: Explicitly stop AFK endpoints on shutdown

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/apple: audio: Create a device link to the DMA device

This works even before the DMA device probes. Might help deal with
runtime-pm ordering, though it doesn't solve the deferred ordering
problem (since we're creating the link while already probing)...

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/apple: audio: Defer DMA channel acquisition to device open

Allow the DMA device driver to probe late, and still create the sound
device upfront. Instead try to request the DMA channel on first PCM
open. This should be safe as long as we bail early and don't allow the
process to continue to configuring buffers (since that requires the DMA
to be configured).

Signed-off-by: Asahi Lina <lina@asahilina.net>

hoshinolina committed:
drm/apple: audio: Fix hotplug notifications

Signed-off-by: Asahi Lina <lina@asahilina.net>

jannau commented on drivers/gpu/drm/apple/apple_drv.c:
```diff
@@ -90,6 +91,39 @@ static int apple_plane_atomic_check(struct drm_plane *plane,
+	 */
+	if ((new_plane_state->crtc_x + 32) > crtc_state->mode.hdisplay ||
+	     (new_plane_state->crtc_y + 32) > crtc_state->mode.vdisplay) {
+		dev_err_once(state->dev->dev,
```
this should be only a single `dev_err_once` use `\n` inside it and concatenate strings.

jannau commented on drivers/gpu/drm/apple/apple_drv.c:
```diff
@@ -90,6 +91,39 @@ static int apple_plane_atomic_check(struct drm_plane *plane,
+	if ((new_plane_state->crtc_x + 32) > crtc_state->mode.hdisplay ||
+	     (new_plane_state->crtc_y + 32) > crtc_state->mode.vdisplay) {
+		dev_err_once(state->dev->dev,
+			"Plane operation would have crashed DCP! Rejected!");
```
log messages shall end with `\n`

chadmed replied:
The `dev_err`s add `\n` themselves. We already have a bunch of extra `\n`s to remove from DCP error messages

jannau replied:
this is not 100% true. printk supports continuation of log statements with without enclosed '\n' (https://www.kernel.org/doc/html/latest/core-api/printk-basics.html#c.pr_cont).

```
pr_info("this is");
pr_cont("one line\n");
```
will produce a single line. So it is generally recommended to include the '\n'.

jannau commented on drivers/gpu/drm/apple/apple_drv.c:
```diff
@@ -269,8 +322,19 @@ static const struct drm_crtc_funcs apple_crtc_funcs = {
+{
+	int ret;
+
+	ret = drm_atomic_normalize_zpos(dev, state);
```
I don't think we should do this. Our handling of the primary plane depends on it being the bottom plane. We should instead use `drm_plane_create_zpos_immutable_property()` to fix the zpos of the planes.

jannau commented on drivers/gpu/drm/apple/apple_drv.c:
```diff
@@ -318,16 +382,29 @@ static int apple_probe_per_dcp(struct device *dev,
 	struct apple_encoder *enc;
-	struct drm_plane *primary;
-	int ret;
+	struct drm_plane *planes[DCP_MAX_PLANES];
```
we only need a single plane pointer (since we do not support a cursor plane)

chadmed replied:
We need the pointers to set the immutable zpos property in a sane way.

jannau commented on drivers/gpu/drm/apple/apple_drv.c:
```diff
@@ -318,16 +382,29 @@ static int apple_probe_per_dcp(struct device *dev,
-	if (IS_ERR(primary))
-		return PTR_ERR(primary);
+	/* Set up our other planes */
+	for (i = 1; i < DCP_MAX_PLANES; i++) {
```
is there any advantage of offering more than overlay plane if we can blend only 2?

chadmed replied:
The only reason I did it like this instead of hard-coding two planes was to make it easier to extend if/when we figure out how to get all four surfaces to play nice somehow. I can revert back to hard coded planes if you don't think there's value in doing it like this

jannau commented on drivers/gpu/drm/apple/iomfb_template.c:
```diff
@@ -1300,7 +1299,19 @@ void DCP_FW_NAME(iomfb_flush)(struct apple_dcp *dcp, struct drm_crtc *crtc, stru
+		 * Plane order is nondeterministic for this iterator. DCP will
+		 * almost always crash at some point if the z order of planes
+		 * flip-flops around. Make sure we are always blending them
+		 * in the correct order.
```
this seems strange and I wouldn't expect that changing the order is the cause of the crash but something else. We of course need a deterministic order for proper display.

chadmed replied:
Could be a race leaving a dangling pointer somewhere from all the surfaces swapping order, but I definitely did isolate the underlying cause to the fact that they were swapping order. If you leave everything as-is and comment out `normalize_zpos` it will happen.

jannau commented on drivers/gpu/drm/apple/iomfb_template.c:
```diff
@@ -1300,7 +1299,19 @@ void DCP_FW_NAME(iomfb_flush)(struct apple_dcp *dcp, struct drm_crtc *crtc, stru
+		 * flip-flops around. Make sure we are always blending them
+		 * in the correct order.
+		 *
+		 * DCP's plane order is the reverse of DRM's, and despite
```
This is not true. surface 0 is below surface 1. My impression of macos behavior:
surface 0 is primary surface supporting only a limited amount of pixel formats
surface 1 is has full pixel format support and is rendered above surface 0 and 2
surface 2 is has full pixel format support

For desktop use macos use surface 0 as primary surface and surface 1 for the cursor
During windowed video playback it uses surface 2 for the video and surface 1 for the visible desktop and SW cursor with a transparent cut-out for the video.

chadmed replied:
Perhaps our ordering of surfaces in incorrect somewhere then, because without reversing the order the cursor is always rendered underneath the desktop (see branch with cursor plane hack)

jannau replied:
The code as is is correct. I was commenting on the comment which misleading/wrong. DCP doesn't z-order surfaces by descending surface index

jannau commented on drivers/gpu/drm/apple/iomfb_template.c:
```diff
@@ -1300,7 +1299,19 @@ void DCP_FW_NAME(iomfb_flush)(struct apple_dcp *dcp, struct drm_crtc *crtc, stru
+		 * having 4 surfaces, we can only blend two. Surface 0 is also
+		 * unusable on some machines, so ignore it.
+		 */
+		l = MAX_BLEND_SURFACES - new_state->normalized_zpos;
```
A better approach with the current state of supported planes and formats would be to hard-code `l` based on plane type

chadmed committed:
drm: apple: make plane zpos immutable

Userspace cannot be trusted to give us a sane zpos value, but given DCP's
requirement that the primary plane always be the bottommost surface, we
can't rely on drm_atomic_normalize_zpos() to do the job for us either.

Make the zpos property immutable, and keep the primary plane at zpos 0.

Signed-off-by: James Calligeros <jcalligeros99@gmail.com>

chadmed commented:
Rebased and addressed feedback

jannau committed:
drm: apple: iomfb: Align buffer size on unmap/free as well

Fixes failure to unmap buffers in dcpep_cb_unmap_piodma() due to the
unaligned size. Further along this causes kernel log splat when DCP
tries to map the buffers again since thye IOVA is still in use.
This causes no apparent issue although map_piodma callback signals an
errror and returns 0 (unmapped as DVA).

It's not clear why this presents only randomly. Possibly some build or
uninitialized memory triggers this unmap/free and immediate allocate/map
cycle in the DCP firmware. I never notices this with a clang-built kernel
on j314c. It showed with gcc build with the Fedora config at least on
6.8.8 based kernels. This did not reproduce on j375d.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: iomfb: Provide the EDID as connector property

External display only since the EDID provided by integrated panels
holds no useful / correct information.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
drm: apple: afk: Optionally match against EPICName

The dpavserv endpoint uses various EPICProviderClass depending on the
connected display.

Observed values:
- "AppleDCPAgileCDIDPDisplay" (j134c, dcp, panel)
- "AppleDCPMCDP29XX" (j274, dcp, hdmi)
- "AppleDCPPS190" (j474s, dcpext0, hdmi)
- "DCPDPService" (j474s, dcpext1, typec)

So match against against EPICName which is consistent in all cases.
This also allows the distinction between 'dcpav-service-epic' and
'dcpdp-service-epic'. Not sure what the second EPIC service is used for.

Signed-off-by: Janne Grunau <j@jannau.net>

jannau committed:
Revert "drm: apple: HACK: Do not delete piodma platform device"

This reverts commit fa86f31f64a691eb65a217c66468b3e9e58cc9e1.

Signed-off-by: Janne Grunau <j@jannau.net>

chadmed commented:
Rebased on top of `bits/200-dcp` and base for PR changed

chadmed commented:
kwin now supports using overlay planes for cursors (queued for 6.3, backported to 6.2 in fedora) so we now have a use case!

jannau commented:
pushed this with some refactoring to bits/200-dcp / asahi-wip
Fixes issues with some igt-gpu-tools tests but overall it's hopeless since igt-gpu-tools assumes a cursor plane is available.
